src/main/java/build/buildfarm/instance/shard/RedisShardSubscription.java

```diff
@@ -88,15 +117,26 @@ private void mainLoop() throws IOException {
   }
 
   public void stop() {
-    if (stopped.compareAndSet(false, true)) {
-      try {
-        subscriber.unsubscribe();
-      } catch (JedisException e) {
-        // subscriber validates with private member to determine connection status
-        // detect this condition and ignore it
-        if (!e.getMessage().endsWith(" is not connected to a Connection.")) {
-          throw e;
+    manageState(SubscriptionAction.STOP);
+    try {
+      subscriber.unsubscribe();
+    } catch (JedisException e) {
+      // If stop() is called before a connection is established, log and throw the exception
+      if (e.getMessage().endsWith(" is not connected to a Connection.")) {
+        log.log(
+            Level.SEVERE,
+            "RedisShardSubscription::stop called but no connection is established. "
+                + "Subscription is now in 'Stopped' state and cannot subscribe.");
+      }
+      throw e;
+    } finally {
```

Comment:
This retry solves the issue that there is no way to tell if the subscription has fully completed. Instead we can tell that the unsubscribe call was successful because `attemptingSubscribe` is now false. The benefit here is that if `subscribe` was attempted, we can guarantee that `stop` will not complete successfully but leave an orphaned subscription. The downside of this change is that `stop` will now block.